src/borrow.rs

```diff
@@ -493,14 +649,59 @@ fn base_address<T, D>(array: &PyArray<T, D>) -> usize {
     }
 }
 
+fn data_range<T, D>(array: &PyArray<T, D>) -> (usize, usize)
+where
+    T: Element,
+    D: Dimension,
+{
+    let data = unsafe { (*array.as_array_ptr()).data };
+
+    let mut start = 0;
+    let mut end = 0;
+
+    for (&dim, &stride) in array.shape().iter().zip(array.strides()) {
+        let offset = dim.saturating_sub(1) as isize * stride;
+
+        if offset >= 0 {
+            end += offset;
+        } else {
+            start += offset;
+        }
+    }
```

Comment:
I noticed that this implementation does not support 0-sized arrays (e.g., shape is `(6, 0, 3)`), but not sure about how to implement it in a smart way.
Also, how about using an exclusive end (i.e., `[start, end)`) so that we can distinguish 0-size and 1-sized array?

Reply:
Switch back to an exclusive data range and handled an array with zero dimensions by checking the shape beforehand. 

Doing it inline of the loop does not really work as I need to do `end += size_of::<T>() as isize;` only if the loop went through. This can be written using a label and breaking, but since labels on blocks are unstable this would require a loop that never loops which feels like it is obfuscating the code to me, e.g.

```rust
#[allow(clippy::never_loop)]
'offsets: loop {
    for (&dim, &stride) in shape.iter().zip(strides) {
        if dim == 0 {
            break 'offsets;
        }

        let offset = (dim - 1) as isize * stride;

        if offset >= 0 {
            end += offset;
        } else {
            start += offset;
        }
    }

    end += size_of::<T>() as isize;
    break 'offsets;
}
```